Expensify/App
# Network: Retry requests that failed in flight

### Details

Add updates to:
- Retry requests when something unexpected caused an error in flight
- Prevent multiple authentication calls running at the same time

### Fixed Issues

$ https://github.com/Expensify/App/issues/5952

### Tests


We're trying to make a request fail in the middle, like connection loss
The easiest way so far has been using Chrome to "slow down the connection" 
On staging it's even enough to cause a timeout (but on localhost I think there are different rules and it doesn't happen)
Check out the web before after video for a thorough demonstration 

#### Web (Chrome) & Desktop
1. Open dev tools and add a very slow connection profile
2. Enable the throttled connection you just made
3. Write a message or two 
  - on staging the request should fail in about 15 sec. just wait
  - on localhost you can kill your router connection now to cause the request to fail
      - you'll probably need to also kill the proxy server and restart it
      - also you might need to go to offline and online (in Chrome) to let it know it really don't have internet..
4. Verify the messages failed (see `Report_AddComment`) in the network tab
5. Set the Network speed to "offline" - now we're acting like we're fully offline
6. Send a few more messages, but do take a note where the message from step 3) are
7. Now set Network speed to "full" and observe the requests getting retried
  - if you tried this on localhost resume you're real internet connection before setting network speed to full in Chrome
8. There should be no pending messages stuck on the bottom 

#### Others
We're trying to do the same as above, but without dev helpers
It sometimes possible to trick the device you're still online, like if you unplug your network cable or use a tethered connection and disable "data" on the tethered connection, then the requests should timeout and fail
When you resume connectivity any pending request should be sent

### QA Steps


See if you can do the "Tests" above, it should be easier on staging and Chrome, just letting the connections timeout using dev tools and throttle
Otherwise a real world case it to use a device in a spotty connection and notice that you're getting offline or messages are sent too slow
Check for regressions around network related stuff like uploading an attachment
Spending some time in airplane mode then resuming
Use the app for a while and monitor for any indication of a regression that might be caused here

### Tested On


### Screenshots


#### Web


##### Before and After (after starts around the 20th second mark)
https://user-images.githubusercontent.com/12156624/144518079-e8face7f-f0dd-44b5-8a2f-950e0828bbc4.mp4

#### Mobile Web

![image](https://user-images.githubusercontent.com/12156624/144520727-9c939100-3ef4-4955-9802-139641329667.png)

#### Desktop

![image](https://user-images.githubusercontent.com/12156624/144520810-cc9a0b3a-fa1f-4b61-82d4-969480a3b57b.png)

#### iOS

![image](https://user-images.githubusercontent.com/12156624/144520636-6c26e88b-4239-46a8-a38d-d31f2ecc40a5.png)

#### Android

https://user-images.githubusercontent.com/12156624/144518247-b11bdb5f-6174-413f-892b-04bf6c57cbe3.mp4
